CORE-11502: pull signing store cache up, clean up interfaces and implementations

dickon opened the pull request:
I reviewed the signing keys parts of the CryptoRepository interface. It turns out that the signing keys were all in the cluster database (since the tenantId was specified as CRYPTO when picking a database), irrespecitve of the tenant ID passed in when calling the SigningService. That's a bug, since virutal node data must be stored in the virtual node database. So, there's no reason for the tenant ID parameter to the signing methods, since an instance of `CryptoRepository` embeds a tenant ID, and so it is ambigious to specify it again. To be clear, this *is* a behaviour change.  

Currently `V1SigningKeyStore` contains both database code and a caching layer, mixed up together. This makes it hard to maintain the database code, and that's pressing since we need to make significant changes. It also means that it is necessary to configure that cache, which brings Corda config into the repository layer. SImplify this code by pushing the caching up to `SigningServiceImpl`. This matches the pattern where the peer code in `SoftCryptoService` does caching and calls down to the repository layer purely for database access. 

A previous attempt to solve this problem involved pulling out some methods to `SigningKeyRepositoryImpl` in order to make those methods unit testable without needing to involve lifecycle.  I believe the root cause of this was the number of OSGi and lifecycle components involved. So, I've removed `SigningServiceFactoryImpl` and so OSGi is no longer involved at this level, so the code is easy to test without needing lifecycle.

dickon committed:
tidy up signing key repo code, split CryptoRepostiory

fowlerrr commented on components/crypto/crypto-persistence-impl/src/main/kotlin/net/corda/crypto/persistence/impl/HSMStoreImpl.kt:
```diff
@@ -84,23 +74,23 @@ class HSMStoreImpl @Activate constructor(
 
     class Impl(
-        private val cryptoRepository: CryptoRepository,
+        private val cryptoHSMRRepository: HSMRepository,
```
typo in `cryptoHSMRepository`

fowlerrr commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoOpsBusProcessor.kt:
```diff
@@ -38,36 +41,50 @@ import net.corda.data.crypto.wire.ops.rpc.commands.SignRpcCommand
+    jpaEntitiesRegistry: JpaEntitiesRegistry,
+    keyEncodingService: KeyEncodingService,
+    virtualNodeInfoReadService: VirtualNodeInfoReadService,
+    layeredPropertyMapFactory: LayeredPropertyMapFactory,
```
I would argue that most (or all) of these are only for creating the `SigningService`, and therefore don't need to be passed in.  Let them remain with the caller.  I know that would interfere with the factory function below, but I think it's better to keep all of that _out_ of this class.

dickon replied:
yeah, much simpler now. I'm not sure how I missed that.

fowlerrr commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoOpsBusServiceImpl.kt:
```diff
@@ -54,7 +75,17 @@ class CryptoOpsBusServiceImpl @Activate constructor(
+            jpaEntitiesRegistry = jpaEntitiesRegistry,
+            virtualNodeInfoReadService = virtualNodeInfoReadService,
+            layeredPropertyMapFactory = layeredPropertyMapFactory
+        )
```
Yeah, I'd say create the `SigningService` here and just pass it in.  It's mockable for tests so there's no need for the factory function.

dickon replied:
done

fowlerrr commented on components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/SigningServiceGeneralTests.kt:
```diff
@@ -394,4 +394,97 @@ class SigningServiceGeneralTests {
+//        }
+//        assertThat(tenantCap.allValues.single()).isEqualTo("tenant")
+//        assertThat(fullIdsCap.allValues.single()).isEqualTo(listOf(hashB.toString()))
+//    }
```
We've already chatted, but just noting that these will come back in some form (in some place)

dickon replied:
What do you think about leaving out the commented out new code in the signing service tests in this PR and then that's another piece either of us can do in parallel with other work? It's not exaclty like we're deleting tests, since that coverage didn't exist at this level

fowlerrr replied:
Yeah, I don't mind coming back to it.  I just wanted to go on record (for others reading this) that we have "a plan"

fowlerrr commented on components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/infra/TestWrappingRepository.kt:
```diff
@@ -0,0 +1,18 @@
+
+    override fun findKey(alias: String): WrappingKeyInfo? = keys[alias]
+
+    override fun close() {
```
meh.  I suppose you could clear the map here.  But it probably doesn't matter much.

dickon replied:
I don't see the point.

fowlerrr commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/WrappingRepository.kt:
```diff
@@ -0,0 +1,9 @@
+import net.corda.crypto.persistence.WrappingKeyInfo
+import java.io.Closeable
+
+interface WrappingRepository : Closeable {
```
Kdocs!!!!

dickon replied:
They didn't exist before.

dickon replied:
done

fowlerrr commented on components/crypto/crypto-softhsm-impl/src/test/kotlin/net/corda/crypto/softhsm/impl/TestSigningRepositoryImpl.kt:
```diff
@@ -20,37 +18,32 @@ import org.mockito.kotlin.times
+            keyEncodingService = mock(),
+            digestService = mock(),
+            layeredPropertyMapFactory = mock()
+        )
```
Er, did you want this again?

fowlerrr commented on components/crypto/crypto-softhsm-impl/src/test/kotlin/net/corda/crypto/softhsm/impl/TestV1SigningRepositoryImpl.kt:
```diff
@@ -13,7 +13,7 @@ import org.mockito.kotlin.mock
 
 
-class TestV1CryptoRepositoryImpl {
+class TestV1SigningRepositoryImpl {
```
Probably don't need the `V1` now

fowlerrr replied:
Also, is this class for testing wrapping keys or signing keys?

fowlerrr commented on gradle.properties:
```diff
@@ -6,6 +6,7 @@ kotlin.stdlib.default.dependency=false
 kotlinMetadataVersion = 0.6.0
 
 org.gradle.jvmargs=-Dfile.encoding=UTF-8
+org.gradle.parallel=true
```
Uncommit!!  Uncommit!!!

fowlerrr reviewed (commented):
I think I prematurely reviewed this and a lot of it would have been fixed anyway, but here are some comments.

dickon committed:
test fixes, simplify construction in CryptoOpsBUsProcessor

dickon committed:
simplify construction in CryptoOpsBUsProcessor

dickon committed:
fix detekt

dickon committed:
add docstrings

dickon committed:
fix typo

dickon committed:
remove duplication

fowlerrr commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoOpsBusServiceImpl.kt:
```diff
@@ -76,15 +78,18 @@ class CryptoOpsBusServiceImpl @Activate constructor(
+                virtualNodeInfoReadService = virtualNodeInfoReadService,
+                layeredPropertyMapFactory = layeredPropertyMapFactory,
+                config = event.config.get(CRYPTO_CONFIG)
+                    ?: throw InvalidParameterException("$CRYPTO_CONFIG missing from config")
```
There's a conveience function for this.
```suggestion
               config = event.config.toCryptoConfig()
```

Different exception but...

dickon committed:
fix detekt

dickon committed:
use convenience function

dickon committed:
fix typo

dickon committed:
remove unused import

dickon committed:
add docstring

dickon committed:
add docstring

dickon committed:
improve name

dickon committed:
port test

dickon committed:
add factory interface for signing repository

dickon committed:
fix test

dickon committed:
fix tests

fowlerrr committed:
Fix up some compile issues

dickon committed:
fix tests

dickon committed:
fix for use of signingRepository factory

dickon committed:
fix detekt

dickon committed:
fix detekt

fowlerrr committed:
Fix tests

dickon committed:
restore test `Should throw original exception failing signing`
remove duplication parameter keyEncodingService

fowlerrr committed:
Fix detekt

fowlerrr committed:
Use CryptoSigningService config as it's more consistent

dickon committed:
remove KeyEncodingService import now we use schema digest directly

dickon committed:
port CryptoOpsBusProcessorTests `Should return empty list for unknown key id`

fowlerrr committed:
Fix more tests

dickon committed:
close EMFs as soon as we can

Do not have close on the test signing repository clear the state

dickon committed:
Reduce EMF lifetime in HSMStore

dickon committed:
use the factory and short lived pattern for wrapping key repository EMF handling

dickon committed:
close EMFs as soon as we can in PersistenceTests.kt

dickon committed:
fix typo

dickon committed:
implement new extension methods for EntityManager and start using them

dickon committed:
mark up the confusing extension as deprecated, clarify kdocs

add a matching consume method to SigningRepository

use the new extension methods in more places

dickon committed:
use my new extension methods for EMF wherever possible, and add similar methods for the *Repository classes

dickon committed:
comment out a couple of NPEing tests for now

dickon committed:
revert the consume method, so I don't have to mock it now

dickon committed:
Revert "comment out a couple of NPEing tests for now"

This reverts commit 6d1fadf524fd7384d962bd337729b688809960a7.

dickon committed:
port further test

add consume function as a helper

dickon committed:
port another test

dickon committed:
revert parallel builds

fowlerrr committed:
Revert EntityManagerUtils change

fowlerrr committed:
Revert change

dickon commented:
I review @fowlerrr's commits to date, and they look good to me.

fowlerrr reviewed (approved):
I've been going over this as we worked on it and it's in good shape to merge for the work that will follow
